<div class="clearfix" style="padding: 10px; padding-left: 0px">
<a href="http://bombora.com"><img src="https://app.box.com/shared/static/e0j9v1xjmubit0inthhgv3llwnoansjp.png" width="200px" class="pull-right" style="display: inline-block; margin: 5px; vertical-align: middle;"></a>
<h1> Bombora Data Science: <br> *Interview Exam* </h1>
</div>

<img width="200px" src="https://app.box.com/shared/static/15slg1mvjd1zldbg3xkj9picjkmhzpa5.png">

---
# Welcome

Welcome! This notebook contains interview exam questions referenced in the *Instructions* section in the `README.md`—please read that first, *before* attempting to answer questions here.

<div class="alert alert-info" role="alert" style="margin: 10px">
<p style="font-weight:bold">ADVICE</p>
<p>*Do not* read these questions, and panic, *before* reading the instructions in `README.md`.</p>
</div>

<div class="alert alert-warning" role="alert" style="margin: 10px">
<p style="font-weight:bold">WARNING</p>

<p>If using <a href="https://try.jupyter.org">try.jupyter.org</a> do not rely on the server for anything you want to last - your server will be <span style="font-weight:bold">deleted after 10 minutes of inactivity</span>. Save often and rember download notebook when you step away (you can always re-upload and start again)!</p>
</div>


## Have fun!

Regardless of outcome, getting to know you is important. Give it your best shot and we'll look forward to following up!

## GITHUB USERNAME: 
ner-aim
## ds-position: 
data-scientist-intern-spring-2022
## mylastname-myfirstname:
Pottapatri,  Pradyumn
## University:
Georgia State University

# Exam Questions

## 1. Algo + Data Structures

### Q 1.1: Fibionacci
![fib image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/200px-Fibonacci_spiral_34.svg.png)

#### Q 1.1.1
Given $n$ where $n \in \mathbb{N}$ (i.e., $n$ is an integer and $n > 0$), write a function `fibonacci(n)` that computes the Fibonacci number $F_n$, where $F_n$ is defined by the recurrence relation:

$$ F_n = F_{n-1} + F_{n-2}$$

with initial conditions of:

$$ F_1 = 1,  F_2 = 1$$

### Answer:

In [14]:
def fibonacci(n):
    if(n<1): return 0
    fib = [1,1]
    p=2
    while(p!=n):
        fib.append(fib[p-1]+fib[p-2])
        p+=1
    return fib[n-1]

In [ ]:
fibonacci

In [16]:
fibonacci(40)/fibonacci(39)

1.6180339887498947

#### Q 1.1.2
What's the complexity of your implementation?

### Answer:

* Since the complexity of the loop is $O(n)$ and there are two operations (append and addition) of $O(1)$, the resulting time complexity is $O(n) + O(1)$ which is $O(n)$. 

* Although, an array is being used to store the values of the series, the space complexity is also $O(n)$.

#### Q 1.1.3
Consider an alternative implementation to compute Fibonacci number $F_n$ and write a new function, `fibonacci2(n)`.

### Answer:

In [19]:
def fibonacci2(n):
    if(n<1): return 0
    a=1
    b=1
    p=2
    while(p!=n):
        c=a+b
        a=b
        b=c
        p+=1
    return c

In [21]:
fibonacci2(40)/fibonacci2(39)

1.6180339887498947

In [22]:
fibonacci2(500)/fibonacci2(499)

1.618033988749895

#### Q 1.1.4
What's the complexity of your implementation?

### Answer:

* The complexity of the above code is $O(n)$ and with 2 iterative and 2 literal operations with $O(1)$ complexity. Hence the time complexity is still $O(n)$.

* Although the space complexity is reduced to $O(1)$ because no extra space is being used. 

#### Q 1.1.5
What are some examples of optimizations that could improve computational performance?


### Answer:

* There are a few methods that can be used to improve the computational performance such as using memoization for better runtime performance. The above methods also have used dynamic programming to calculate the values of fibonnaci series. These methods can calculate fibonacci series with a time complexity of $O(n)$ and space complexity of $O(1)$. Although the memoization has a runtime complexity of $O(log_en)$ using Fibonacci matrix corelation, it also has a space complexity of $O(log_en)$. 

* The best possible way for better complexity is by using Golden ratio $\phi$ using Benet's formula of $Fibonacci(n) = \bigg[\frac{(\phi)^n - (-\phi)^n}{\sqrt{5}}\bigg]$

* The above has a runtime complexity of $O(log_en)$ because exponent takes $log_e n$ time to calculate with no extra space to afford i.e. we can make it work in $O(1)$ space.

PS: An easier way of checking the fibonacci series is using the Golden ratio because the $\frac{F_n}{F_{n-1}}$ ~  $\phi$ for most of the time. 

### Q 1.2: Linked List
![ll img](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Singly-linked-list.svg/500px-Singly-linked-list.svg.png)

#### Q 1.2.1
Consider a [singly linked list](https://en.wikipedia.org/wiki/Linked_list), $L$. Write a function `is_palindrome(L)` that detects if $L$ is a [palindrome](https://en.wikipedia.org/wiki/Palindrome), by returning a bool, `True` or `False`.


#### Q 1.2.2
What is the complexity of your implementation?

#### Q 1.2.3
Consider an alternative implementation to detect if L is a palindrome and write a new function, `is_palindrome2(L)`.

#### Q 1.2.4
What's the complexity of this implementation?


#### Q 1.2.5 
What are some examples of optimizations that could improve computational performance?


## 2. Prob + Stats

### Q 2.1: Finding $\pi$ in a random uniform?
<img src=https://www.epicurus.com/food/recipes/wp-content/uploads/2015/03/Pi-Day.jpg width="480">

Given a uniform random generator $[0,1)$ (e.g., use your language's standard libary to generate random value), write a a function `compute_pi` to compute [$\pi$](https://en.wikipedia.org/wiki/Pi).

#### Answer:

Using Monte Carlo Method to estimate Pi in a square. Monte Carlo says that number of points for a random object inside a rectangle can give the area of the random object. (From Modelling and Simulation in undergraduate)

To calculate the area of a circle, we take a square of unit length where the circle's diameter is the line connecting the midpoints of two opposite sides. The ratio of the area of the circle to the square is given by:

$$\frac{\text{Points covering the area of circle}}{\text{Points covering the area of square}} = \frac{\pi*(0.5)^2}{1^2} = \frac{\pi}{4}$$

$$\implies \pi = 4*\frac{\text{Points covering the area of circle}}{\text{Points covering the area of square}}$$

In [44]:

import random as rd
def compute_pi(n):
    c = 0
    s = 0
    for i in range(n^2):
        x = rd.uniform(0,1)
        y = rd.uniform(0,1)
        dist = (x**2+y**2)
        if(dist<=1): c+=1
        s+=1
    return 4*c/s

In [66]:
compute_pi(10000)

3.123375324935013

The above is a uniform random number generator which estimates Pi. 

### Q 2.2: Making a 6-side die roll a 7?

Using a single 6-side die, how can you generate a random number between 1 - 7?

### Q 2.3: Is normality uniform?

<img src=https://rednaxela1618.files.wordpress.com/2014/06/uniformnormal.png width="480">


Given draws from a normal distribution with known parameters, how can you simulate draws from a uniform distribution?

### Q 2.4: Should you pay or should you go?

![coin flip](https://lh5.ggpht.com/iwD6MnHeHVAXNBgrO7r4N9MQxxYi6wT9vb0Mqu905zTnNlBciONAA98BqafyjzC06Q=w300)

Let’s say we play a game where I keep flipping a coin until I get heads. If the first time I get heads is on the nth coin, then I pay you $2^{(n-1)}$ US dollars. How much would you pay me to play this game? Explain.

### Q 2.5: Uber vs. Lyft

![uber vs lyft](http://usiaffinity.typepad.com/.a/6a01347fc1cb08970c01bb0876bcbe970d-pi)

You request 2 UberX’s and 3 Lyfts. If the time that each takes to reach you is IID, what is the probability that all the Lyfts arrive first? What is the probability that all the UberX’s arrive first?

### Q 2.6: Pick your prize
<img src=https://miro.medium.com/max/1100/1*m5b3O9sE68UCXjLw5oxy2g.png width="480">

A prize is placed at random behind one of three doors and you are asked to pick a door. To be concrete, say you always pick door 1. Now the game host chooses one of door 2 or 3, opens it and shows you that it is empty. They then give you the option to keep your picked door or switch to the unopened door. Should you stay or switch if you want to maximize your probability of winning the prize?

## 3 Conceptual ML

### Q 3.1 Why study gradient boosting or neural networks?

Consider a regression setting where $X \in \mathbb{R}^p$ and $Y \in \mathbb{R}$. The goal is to come up with a function $f(X): \mathbb{R}^p \rightarrow \mathbb{R}$ that minimizes the squared-error loss $(Y - f(X))^2$. Since X, Y are random variables, we seek to minimize the expectation of the squared error loss as follows
\begin{equation}
EPE(f) = \mathbb{E}\left[(Y-f(X)^2\right]
\end{equation}
where EPE stands for expected prediction error. One can show that minimizing the expected prediction error leads to the following _regression function_
\begin{equation}
f(x) = \mathbb{E}\left[Y|X=x\right]
\end{equation}

The goal of any method is to approximate the regression function above, which we denote as $\hat{f}(x)$. For example, linear regression explicitly assumes that the regression function is approximately linear in its arguments, i.e. $\hat{f}(x) = x^T\beta$ while a neural network provides a nonlinear approximation of the regression function. 

The simplest of all these methods is [k-nearest neighbors](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm). Given $x$ and some neighbourhood of $k$ points $N_k(x)$, $\hat{f}(x)$ is simply the average of all $y_i|x_i \in N_k(x)$.  Let $N$ denote the training sample size. Under mild regularity conditions on the joint probability distribution $Pr(X, Y)$, one can show that as $N \rightarrow \infty$, $k \rightarrow \infty$ such that $k/N \rightarrow 0$, then $\hat{f}(x) \rightarrow f(x)$ where $\rightarrow$ means approaches or goes to. In other words, the k-nearest neighbors algorithm converges to the ideal solution as both the training sample size and number of neighbors increase to infinity.

Now given this _universal approximator_, why look any further and research other methods? Please share your thoughts.


### Q 3.2 Model Selection and Assesment

Consider a multiclass classification problem with a large number of features $p >> N$, for e.g $p=10000, N=100$ The task is threefold
1. Find a "good" subset of features that show strong _univariate_ correlation with class labels
2. Using the "good" subset, build a multi class classifier
3. Estimate the generalization error of the final model

Given this dataset, outline your approach and please be sure to cover the following
- Data splitting
- Model Selection: either estimating the performance of different classifiers or the same classifier with different hyperparameters
- Model Assessment: having chosen a classifier, estimating the generalization error

Assume all features are numerical, the dataset contains no NULLS, outliers, etc. and doesn't require any preprocessing.



#### Answer:

* 
##### Step 1: 
    Since the number of features are way higher than that of the class, each of the features needs to be determined with the response variable. For this univariate analysis, there are a number of methods that can be used. The easiest and direct way for the procedure is using Pearson's Correlation to generate high and low noise of the feature and comparing with itself. We compare the p-values of the features and generate correlations with respect to the response variable. Although this works for a linear relationship between the variables. This fails for non-linear variables. 
    <br><br>
    
    This can be avoided using the Distance correlation as a counter measure to address the shortcomings of the Pearson's model. Pearson's correlation works well with smaller datasets, but assuming this is a medoicre dataset and all the features are numerical, we can say that Distance Correlation will do a great job. 
    
* 
##### Step 2:
    Now that we have a good subset we need to split the data into train and test. The normal way would be using 70% and 30% ratio of the splitting the dataset. Although this can cause further problems, like overfitting of bias towards a single kind of label. To ensure the data has been split into proper proportions, we require the data split to be uniform and have class distributions similar to the process of handling Big Data. So we split the data into Train, Validation and Test set. As usual, we train the dataset and outline the work by using the validation set at each epoch or Cross validation to determine the model performance. Later on, the testing set is used to determine the nature of the model.
    <br><br>
    Since we are using the validation set, a bunch of multiclassification models can be used to determine the best for the task. As a general rule of thumb, because all the features are numeric, tree models can be used to determine the task, assuming there is no nulls, outliers and the sample is perfectly balanced. Regardless of the sample being balanced, which is generally not the case with multiclass classification. So we can use oversampling for training the dataset. Training and validation can be done be using the tree models and Boosting models such as Decision Tree, Random Forest, XGBoosting etc. 
    
*
##### Step 3:
    Now that we have a good classifier using the hyperparameters working on the good validation set, we need to test the model on the testing set. If the validation error is low i.e. the model performs well on the training data, and the generalization error is high, the model is overfitting on the training dataset. To avoid that, we need to get the hyperparameters properly and assign it to the testing data. We can calculate the generalization error as $E_gen = \int error(f_d(x),y)p(y,x)dx$, p(y,x) is how often we see y,x. This is the optimistic error generalization. We compare it with different cross-validation of different algorithms.